### ISSUE TITLE
Not able to connect via WebUsbTransport

### ISSUE THREAD

**Issue by resetko-corp** [trezor/trezor-suite#9644]
**Describe the bug**
Hey. Trying to set up an example app using the latest version of `@trezor/connect-web`, but have issues connecting through webusb transport.

**Info:**

-   Remembered Wallet: yes/no (not sure about this one)
-   Suite Version: `@trezor/connect-web` 9.1.4
-   Browser: `Brave v1.59.117` AND `Chrome 117.0.5938.149`
-   OS: Linux, 6.5.6-zen2-1-zen
-   Bridge: no
-   Firmware Version: Model T, https://github.com/trezor/trezor-firmware/blob/88e1f8c7a5c7615723664c64b0a25adc0c409dee/core/CHANGELOG.md

**How to reproduce**

My setup is fairly simple app, consists of HTML file with TS file compiled by Vite:

```import TrezorConnect from "@trezor/connect-web";

const keysButton = document.getElementById("keys-button")!;
const initButton = document.getElementById("init-button")!;

initButton.addEventListener("click", () => {
  TrezorConnect.init({
    lazyLoad: true, // this param will prevent iframe injection until TrezorConnect.method will be called
    manifest: {
      email: "developer@xyz.com",
      appUrl: "http://your.application.com",
    },
    transports: ["WebUsbTransport"],
  })
    .then(() => {
      console.log("trezor init done");
      keysButton.removeAttribute("disabled");
    })
    .catch((e: unknown) => {
      console.error("trezor init error", e);
    });
});

keysButton.addEventListener("click", () => {
  TrezorConnect.getPublicKey({
    coin: "ETH",
    path: `m/44'/60'/0'/0`,
  })
    .then((response: any) => {
      if (!response.success) {
        throw response.payload;
      }
      console.log("trezor getPublicKey response", response);
    })
    .catch((e) => {
      console.log("trezor getPublicKey error", e);
    });
});
```

So I first do init, when init is done I try to pull public key.

It fails with error `{error: 'Transport is missing', code: 'Transport_Missing'}` and opens a popup with "Install bridge".

I'm observing same behavior when using https://trezor.github.io/trezor-suite/connect-explorer/#/method/ethereumGetPublicKey

Honestly saying I'm not really comfortable installing software on each device where I use Trezor, and assuming we have webusb transport mentioned in `init` method [documentation](https://github.com/trezor/trezor-suite/blob/8f03ebe061fd29c57e8d0fb0e4cfc776dfcb636b/docs/packages/connect/methods/init.md) as one of the options, but not the only one, I thought I can keep using WebUsb if I don't want to install Bridge. Or my understanding is wrong?

At the same time, https://suite.trezor.io/web/ works flawlessly (able to connect and communicate with the device)

**Expected behavior**
Able to connect though webusb without Bridge, same as on https://suite.trezor.io/web/

**Comment by resetko-corp** [comment]
https://github.com/trezor/trezor-suite/assets/112621289/b87d4f65-5ee4-4d31-9ce5-309db99748c6


https://github.com/trezor/trezor-suite/assets/112621289/6bbbe90b-c883-4a55-aa86-9a87b8861442


https://github.com/trezor/trezor-suite/assets/112621289/ae9efac6-9b8a-48b9-ad2d-95f8d239a2f2

**Comment by Hannsek** [comment]
Which browser do you use?

**Comment by resetko-corp** [comment]
> Which browser do you use?

I'm using `Brave v1.59.117` and `Chrome 117.0.5938.149` with same results

**Comment by mroz22** [comment]
when you are bulding an app running on diffrent domain than iframe is loaded from, webusb unfortunatelly does not work due to some chrome restrictions. but maybe, we could render a button in an iframe. let me check if this is feasible.

**Comment by resetko-corp** [comment]
> when you are bulding an app running on diffrent domain than iframe is loaded from, webusb unfortunatelly does not work due to some chrome restrictions. but maybe, we could render a button in an iframe. let me check if this is feasible.

Hm, I'm a bit lost. Indeed I have some problems with iframes in some other setup, but in this particular issue there is no iframe (yet), at least from the side of my test app. Which iframe you're referring to?

**Comment by Hannsek** [comment]
Correct me if I am wrong @mroz22, but he is referring to iframe that is loaded from trezor.io domain when some Connect method (here `TrezorConnect.getPublicKey`) is called.

**Comment by mroz22** [comment]
In the next connect release https://github.com/trezor/trezor-suite/pull/9672 it will become possible to use webusb on localhost. Looking now if it will be possible to make it work also for production

**Comment by resetko-corp** [comment]
Hope will help :crossed_fingers:, because honestly I'm not sure if I can triage the issue. Here are the experiments I'm doing trying to find the root cause - there is a small [repo with test app](https://github.com/resetko-zeal/web-usb-trezor-issues) (you can pull it, do `yarn && yarn dev`, I'm using yarn berry with node20)

So there is the simple app I've mentioned above, now I'm using 3 buttons - first is to init trezor, second to fetch public keys. Also, I added the button to fetch trezor device via WebUSB. In addition, I have set up https, since it's a requirement for all policies to properly apply. And the final thing - I've added locally 3 items to `/etc/hosts` so it's not just localhost:

```
127.0.0.1	foo-example.com
127.0.0.1	bar-example.com
127.0.0.1	baz-example.com
```

You probably would need to visit all 3 URLs and accept a self-signed certificate for it to work within an iframe.

And this is actually close to the structure of our app - we have some page, we have an iframe within it, and we call trezor from within this iframe.

Interesting fact that in my setup, webusb works in all 3 nested frames, despite the fact they all cross-origin.

Attaching video of how I testing it.

https://github.com/trezor/trezor-suite/assets/112621289/903f132d-9b9f-4a8c-b365-c8de307dfe8d

**Comment by Hannsek** [comment]
> Interesting fact that in my setup, webusb works in all 3 nested frames, despite the fact they all cross-origin.

I don't think webusb worked in any of these frames as you needed to install bridge.

**Comment by resetko-corp** [comment]
> > Interesting fact that in my setup, webusb works in all 3 nested frames, despite the fact they all cross-origin.
> 
> I don't think webusb worked in any of these frames as you needed to install bridge.

I mean the JavaScript WebUSB API itself works, because I was able to enumerate devices and open trezor device via `await device.open();` so I'm not sure if Chrome itself blocks webusb API for cross-origin frames, at least I wasn't able to reproduce any problems with browser API for cross-origin frames

**Comment by mroz22** [comment]
ciao, 
we are working on solution to make webusb available. it will work like this:
- call TrezorConnect.something()
- we detect that you don't have bridge
- npm connect package renders button in some overlay overlay which calls navigator.usb.requestDevice
- native browser dialogue opens, your page is granted permisions and this permision is now available to connect-iframe which can now enumerate devices
- npm package opens connect-popup and you continue as you were used to.

**Comment by resetko-corp** [comment]
Hey, I bet when https://github.com/trezor/trezor-suite/pull/9755 is getting release it should be fixed, am I correct?

**Comment by mroz22** [comment]
nope, but likely in the release after this one yes

**Comment by graadient** [comment]
I also can't connect on your reference implementation here and just get a spinning loader (latest Chrome & latest Suite): https://trezor.github.io/connect-explorer/#/method/ethereumGetPublicKey

But most concerning is that **it completely stops the Trezor from working in my wallet**. Was hoping the release 3 days ago would fix it but it didn't. Not being able to use the Trezor for base use cases is quite critical. Any estimate when this will be fixed so the Trezor works again? Thanks!

**Comment by Hannsek** [comment]
Can you elaborate more about _it completely stops the Trezor from working in my wallet_? What happened? How does it behave? What is the name and version of the wallet? …

**Comment by mroz22** [comment]
> https://trezor.github.io/connect-explorer/#/method/ethereumGetPublicKey 

hey this is connect v8 we should put it down. this is no longer actively mantained. try using https://connect.trezor.io/9/#/

**Comment by graadient** [comment]
> Can you elaborate more about _it completely stops the Trezor from working in my wallet_? What happened? How does it behave? What is the name and version of the wallet? …

When trying to access the Trezor's addresses or queue up a transaction on any web3 dapp with an address I have added before, it just gets stuck in a "Loading ..." screen with a spinning green blob around it. It's been working for more than half a year and then stopped a few weeks ago.

This is on [Zeal wallet v3.44](https://chrome.google.com/webstore/detail/zeal-wallet/heamnjbnflcikcggoiplibfommfbkjpj)

**Comment by Hannsek** [comment]
Oh, so you mean Connect. Most likely related to https://github.com/trezor/trezor-suite/issues/4878 and https://github.com/trezor/trezor-suite/issues/6129.

**Comment by graadient** [comment]
[Zeal wallet v3.47](https://chromewebstore.google.com/detail/zeal-wallet/heamnjbnflcikcggoiplibfommfbkjpj) updated to Trezor connect v9 but it still doesn't work in Chrome.

**Comment by Hannsek** [comment]
I can confirm, I am stuck on this page:

![Screenshot 2023-12-07 at 17 04 47](https://github.com/trezor/trezor-suite/assets/57008159/f712a3b2-5cd6-4786-a3ef-77608ea91993)

I can see that TREZOR is paired, however the name of the actual Trezor is not TREZOR.
![Screenshot 2023-12-07 at 17 05 31](https://github.com/trezor/trezor-suite/assets/57008159/f8346d87-c4a7-462d-afeb-14d42d03471d)

If I reload the page it is blank
![Screenshot 2023-12-07 at 17 05 01](https://github.com/trezor/trezor-suite/assets/57008159/6df4d9c0-287b-41fb-94dd-aba14243f55c)


@trezor/qa please also check.

**Comment by Hannsek** [comment]
In the console, I can see this error: 
![image](https://github.com/trezor/trezor-suite/assets/57008159/8deec9f8-fd62-46d3-b96b-53e500d8b394)

**Comment by mroz22** [comment]
I can see that you are probably calling TrezorConnect from withing iframe. Which probably results into TrezorConnect injecting its own iframe inside your iframe. 

![image](https://github.com/trezor/trezor-suite/assets/30367552/aae14875-9f6e-4b6c-8a65-e2cdefff8bf6)

I could imagine, that under this situation popup.html is not able to connect to the nested trezor-connect iframe.

You may now try to:
1. not running this code from iframe
2. consider using [connect-webextension package](https://github.com/trezor/trezor-suite/blob/develop/packages/connect-webextension/README.md) but this would require you to implement connect inside service worker. Also this package is not yet in npm but should appear there within couple of days.
3. wait for me to check if this could be fixed easily. Maybe fix won't be hard, just somehow ensure that popup.html is able to find the right nested iframe. But I am little suspicious if this is going to work. You are saying that it did work before and stopped now? And it still works in firefox? Thats gonna be that chrome version 115 update I think in that case
4. wait for me to implement fallback. In case detect that communication with connect core inside injected iframe is not possible, we are going to load connect core into popup and do all the magic there. It will have slithly less feature (no events, etc), but it should work for the basic cases. Can't really give you any eta now.

**Comment by mroz22** [comment]
ad 2. https://www.npmjs.com/package/@trezor/connect-webextension

**Comment by martykan** [comment]
Should be resolved by the new coreMode in Connect [9.3.1-beta.1](https://connect.trezor.io/9.3.1-beta.1/changelog/)

**Comment by evgenysl** [comment]
Checked webUSB in Chrome and Brave.
Zeal extension also works correctly.

**QA OK**
